packages/libs/sdk/src/solana/solana.ts

```diff
@@ -0,0 +1,168 @@
+// eslint-disable-next-line @nx/enforce-module-boundaries
+import {
+  ComputeBudgetProgram,
+  Connection,
+  TransactionInstruction,
+  PublicKey,
+  TransactionMessage,
+  VersionedTransaction,
+} from '@solana/web3.js';
+import {
+  type EstimatePriorityFeesParams,
+  type PriorityFeeRequestPayload,
+  type PriorityFeeResponseData,
+  type PriorityFeeLevels,
+  type SolanaClientArgs,
+  type SendSmartTransactionArgs,
+  type PrepareSmartTransactionArgs,
+} from './types';
+
+export class Solana {
+  readonly endpointUrl: string;
+  readonly connection: Connection;
+
+  constructor({ endpointUrl }: SolanaClientArgs) {
+    this.endpointUrl = endpointUrl;
+    this.connection = new Connection(endpointUrl);
+  }
+
+  /**
+   * Sends a transaction with a dynamically generated priority fee based on the current network conditions and compute units needed by the transaction.
+   */
+  async sendSmartTransaction(args: SendSmartTransactionArgs) {
+    const {
+      transaction,
+      keyPair,
+      feeLevel = 'medium',
+      sendTransactionOptions = {},
+    } = args;
+    const smartTransaction = await this.prepareSmartTransaction({
+      transaction,
+      payerPublicKey: keyPair.publicKey,
+      feeLevel,
+    });
+    smartTransaction.sign(keyPair);
+
+    const hash = await this.connection.sendRawTransaction(
+      transaction.serialize(),
+      { skipPreflight: true, ...sendTransactionOptions }
+    );
+
+    return hash;
+  }
+
+  /**
+   * Prepares a transaction to be sent with a dynamically generated priority fee based
+   * on the current network conditions. It adds a `setComputeUnitPrice` instruction to the transaction
+   * and simulates the transaction to estimate the number of compute units it will consume.
+   * The returned transaction still needs to be signed and sent to the network.
+   */
+  async prepareSmartTransaction(args: PrepareSmartTransactionArgs) {
+    const { transaction, payerPublicKey, feeLevel = 'medium' } = args;
+    const computeUnitPriceInstruction =
```

Comment:
is there any risk that these values go stale while we process the rest of the function?